### ISSUE TITLE
Support an option that would not use canvas if an image exceeds canvas support

### ISSUE THREAD

**Issue by mejackreed** [openseadragon/openseadragon#1500]
For images that exceed the bounds of `<canvas>` HTML support, would OSD consider an option that would ignore a `useCanvas` option that is specified.

I'm thinking specifically for uses where an OpenSeaDragon viewer is instantiated and then perhaps later on a tileSource is added  which requests images that exceed the bounds available by canvas.

Right now, an OSD viewer canvas will disappear when these pixels are being added to the canvas.

With a simple image source: https://stacks.stanford.edu/image/iiif/mx460bt7026%252Fmx460bt7026_0001/0,0,2048,220740/512,/0/default.jpg

Or with a IIIF image tileSource (zooming past level 3):
https://stacks.stanford.edu/image/iiif/mx460bt7026%2Fmx460bt7026_0001/info.json

Or is the another approach where OSD could be more clever with this large images?

Non authoritative old answer for Chrome bounds: https://stackoverflow.com/a/11585939
```
Maximum height/width: 32,767 pixels
Maximum area: 268,435,456 pixels (e.g., 16,384 x 16,384)
```

**Comment by iangilman** [comment]
Hmm. I wonder how OSD does if you try and switch from canvas to HTML rendering mid-run. Seems like things would go more smoothly if the developer knew ahead of time that they were going to be displaying giant tiles and turned canvas off at the beginning. 

Anyway, it would be interesting to explore better ways to deal with such images. For instance, right now we have a single canvas per tile. Perhaps for large tiles we could have multiple canvases? For that matter, would it work if we skipped copying into a canvas and draw directly from the image into the viewer canvas? I don't recall why we went to storing all of the tiles in canvases. 

Another possible way to deal with this is to see whether we can mix the canvas and HTML renderers, so all of the non-giant tiles can still be rendered via canvas, but the giant tiles can be rendered via HTML. 

Anyway, it would be great to have some sort of solution. I guess I'd like to see if we can fix the canvas renderer first, since that will likely get us the best quality. 

Are you interested in working on this?

**Comment by mejackreed** [comment]
Yep I am. If there is a helpful direction you could point me in, I'd be willing to give this a stab. Does looking at fixing the canvas renderer first seem to be the logical starting point?

**Comment by iangilman** [comment]
Fantastic! And yes, I think the first thing to try is making it so the canvas renderer doesn't need to copy tiles into their own canvas before drawing. I suppose that also means validating that drawing from a giant `img` straight into canvas doesn't have the same size restriction (since the canvas itself isn't that big). 

Here's where the tile canvas is used: 

https://github.com/openseadragon/openseadragon/blob/master/src/tile.js#L380

... So it's a matter of working backward from there, I suppose. Again, I'm not entirely sure why we always make a canvas instead of sometimes using the `img` directly. We certainly want to make canvases for when the user wants to modify the tiles before drawing, so it may be that we just do it always just in case. We still want to enable the modification of tiles, so maybe we'll need a "I don't need to modify tiles" flag for this new mode.

It looks like the thing that generates the tile canvas is `getRenderedContext`. Search the code for everywhere that's called, and provide an alternate route. Note that `getImage` is the counterpart that returns an `img`. Also note that once the rendered context exists, the `img` is released. 

You should also be aware of `context2d`, which is the place that modified tiles get stored. We just don't want to break that pathway.

Okay, I chased down where we added the "always copy to canvas": 

https://github.com/openseadragon/openseadragon/commit/2fc6cc3876b1686f2d08b9089981b2897263a87a

Looks like it's for the performance optimization "Pre-render to an off-screen canvas" mentioned in https://www.html5rocks.com/en/tutorials/canvas/performance/ . I don't know if that optimization is still relevant with today's browsers, but I wouldn't be surprised if it was. That's another reason for this new "don't copy" path to be an explicit flag. I guess the other option would be for it to be automatic, but only for giant tiles.

Anyway, hopefully this gives you some good info to get started with! Let me know how it goes :)

**Comment by mejackreed** [comment]
++ Thanks @iangilman

**Comment by mejackreed** [comment]
Haven't been able to make much progress on this because of other concerns. If anyone else wants to take a shot at this please do.

**Comment by iangilman** [comment]
Thanks for the update... Hopefully someone will!
